### PR TITLE
Hit tables and more

### PR DESCRIPTION
**Pull request by vigo2**
This is an interim PR, before it get's too large ;)

Modelling-wise, all yellow attacks use two-roll hit tables now.

There's a few related bug fixes (e.g. split main/off-hand hit tables for racial weapon expertise), and generate_hit() has been refactored to use new-style Hit_tables. swing_weapon() and whirlwind() no longer use Hit_outcome-vectors, which is a substantial speed up for debug builds (30-40%).

Rampage also got a more impactful bug fix (it was essentially missing a stack), and its ramp-up is modelled more accurately.

The Buff_manager received some "drive-by" bug fixes, but I think that's the first component that needs a full rewrite, anyway.

Update: bonus bug fix for Mace Specialization ;)

### CODE REVIEW

File: sim_interface/source/sim_interface.cpp [hunk 1]
```diff
@@ -857,7 +845,7 @@ Sim_output Sim_interface::simulate(const Sim_input& input)
         compute_talent_weight(simulator_talent, character, talents_info, "Improved Overpower", config,
                               &Combat_simulator_config::talents_t::overpower, 2);
 
-        if (config.combat.use_slam && is_two_handed)
+        if (config.combat.use_slam && is_dual_wield)
```
Comment (TheGroxEmpire) [inline sim_interface/source/sim_interface.cpp:848]: Can you elaborate on this and several ``is_two_handed`` that has been replaced with ``is_dual_wield``?
Wouldn't this meant that it'll compute slam talent if you are dual wielding?

File: simulator/source/Combat_simulator.cpp [hunk 2]
```diff
@@ -256,115 +232,33 @@ void Combat_simulator::cout_damage_parse(Combat_simulator::Hit_type hit_type, So
...
-        }
-        cout_damage_parse(hit_type, weapon_hand, hit_outcome);
-    }
+    auto hit_outcome = hit_table.generate_hit(damage);
+
+    cout_damage_parse(weapon, hit_table, hit_outcome);
 
     if (sweeping_strikes_charges_ > 0)
     {
+        // TODO can_sweep is actually either "clone damage" (can't crit; fully dependent on hit_outcome)
```
Comment (TheGroxEmpire) [inline simulator/source/Combat_simulator.cpp:257]: I've seen sweeping strikes critting on classic vanilla, it uses spell crit chance and modifier (x1.5). Although it is pretty minor and doesn't affect much in grand scheme to be worth implementing

**Review by TheGroxEmpire — CHANGES_REQUESTED** [review]
> Modelling-wise, all yellow attacks use two-roll hit tables now.

Some spells like WW and MS that uses weapon damage can't be a critical block. Quoting from the result that Magey posted in FC discord:
> Category 1 - melee spells that can't be a critical block
this seems to consist of mostly spells that have a weapon damage component so think mortal strike, sinister strike, whirlwind, heroic strike, etc.
any spell that falls into this category rolls block in the first roll and crit in the second roll, meaning their crit chance factor also includes block and so their paperdoll crit penalty is higher than spells that roll block in the second roll

Would this account for that too, or these spells also suffers from double roll crit penalty?

File: simulator/source/Combat_simulator.cpp (hunk 2 above)
Comment (vigo2) [reply simulator/source/Combat_simulator.cpp:257]: No more, it seems - I've tested myself, pretty extensively, and looked through a couple of logs. On "Warcraft Logs", you can find https://tbc.wowhead.com/spell=12723/sweeping-strikes ("clone damage"), and https://tbc.wowhead.com/spell=26654/sweeping-strikes ("normalized swing", for execute & whirlwind). Both are marked as "Can't crit".

**Comment by vigo2** [comment]
> Would this account for that too, or these spells also suffers from double roll crit penalty?

Yes, they all do. Category 1 (and 3) have a (100% - miss% - dodge% - parry% - block%) as their "crit chance factor", Category 2 is using (100% - miss% - dodge% - parry%), but block% x crit% are blocked crits.

Without parry% and block%, they're all using (100% - miss% - dodge%) as "crit chance factor".

File: sim_interface/source/sim_interface.cpp (hunk 1 above)
Comment (vigo2) [reply sim_interface/source/sim_interface.cpp:848]: Good catch. I think this is the fallout of an incomplete refactoring. Looking into it.

Comment (vigo2) [reply sim_interface/source/sim_interface.cpp]: Ye, I noticed that `is_two_handed` was `!character.is_dual_wield()`, and planned to invert it everywhere (as the first couple of usages where `!is_two_handed`). Obviously I stopped after a few lines, and forgot about it.

**Review by TheGroxEmpire — CHANGES_REQUESTED** [review]
Looks like there's an error on how glancing damage is displayed on the sim
image

**Comment by vigo2** [comment]
<Ping> ;)

**Comment by TheGroxEmpire** [comment]
> <Ping> ;)

Sorry for the holdout. The changes look good to me! I'll proceed ahead and merge this with the master. Good work on this, your contributions to the sim has been substantial.

**Pull request merged**